### PR TITLE
Non const iterator neighbor list

### PR DESCRIPTION
**Pull request by atmyers**
Iterate over neighbors in both `const` and non-`const` fashion.

### CODE REVIEW

File: Src/Particle/AMReX_NeighborList.H [hunk 1]
```diff
@@ -10,77 +10,127 @@ namespace amrex
...
         bool operator!= (iterator const& rhs) const { return m_index < m_stop; }
-        
+
         AMREX_GPU_HOST_DEVICE
-        ParticleType operator* () const { return m_pstruct[m_nbor_list_ptr[m_index]];  }
-        
+        const ParticleType& operator* () const { return m_pstruct[m_nbor_list_ptr[m_index]];  }
+
+        AMREX_GPU_HOST_DEVICE
+        ParticleType& operator* () { return m_pstruct[m_nbor_list_ptr[m_index]];  }
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_NeighborList.H]: I think there should be only one version `ParticleType& operator* () const`.  `Iterator const` should still return non-const reference.  User could assign to to `const&` if they want.

See https://en.cppreference.com/w/cpp/iterator/iterator

Comment (atmyers) [reply Src/Particle/AMReX_NeighborList.H]: Okay, done.

File: Src/Particle/AMReX_NeighborList.H [hunk 2]
```diff
@@ -10,77 +10,124 @@ namespace amrex
...
         AMREX_GPU_HOST_DEVICE
         void operator++ () { ++m_index;; }
-        
+
         AMREX_GPU_HOST_DEVICE
         bool operator!= (iterator const& rhs) const { return m_index < m_stop; }
-        
+
+        AMREX_GPU_HOST_DEVICE
+        ParticleType& operator* () { return m_pstruct[m_nbor_list_ptr[m_index]];  }
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
        ParticleType& operator* () const { return m_pstruct[m_nbor_list_ptr[m_index]];  }
```

**Review by WeiqunZhang — APPROVED** [review]

**Review by JBlaschke — APPROVED** [review]
All checks out from my end

**Comment by atmyers** [comment]
Thanks!

**Comment by JBlaschke** [comment]
I just ran 2 FHDeX execs (both slightly different uses of immersed boundaries) on CPU -- looks fine from my end @atmyers

**Comment by JBlaschke** [comment]
+1 🍺

**Pull request merged**